pkg/remote/crd_upgrade.go

```diff
@@ -21,6 +25,136 @@ func PatchCRD(ctx context.Context, clnt client.Client, crd *v1extensions.CustomR
 		client.FieldOwner(v1beta2.OperatorName))
 }
 
+type CrdType string
+
+const (
+	KCP CrdType = "KCP"
+	SKR CrdType = "SKR"
+)
+
+func updateRemoteCRD(ctx context.Context, kyma *v1beta2.Kyma, runtimeClient Client,
+	crdFromRuntime *v1extensions.CustomResourceDefinition, kcpCrd *v1extensions.CustomResourceDefinition) (bool, error) {
+	if ShouldPatchRemoteCRD(crdFromRuntime, kcpCrd, kyma) {
+		err := PatchCRD(ctx, runtimeClient, kcpCrd)
+		if err != nil {
+			return false, err
+		}
+
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func ShouldPatchRemoteCRD(
+	runtimeCrd *v1extensions.CustomResourceDefinition, kcpCrd *v1extensions.CustomResourceDefinition,
+	kyma *v1beta2.Kyma) bool {
+	if runtimeCrd == nil {
+		return true
+	}
+	kcpAnnotation := getAnnotation(kcpCrd, KCP)
+	skrAnnotation := getAnnotation(runtimeCrd, SKR)
+
+	latestGeneration := strconv.FormatInt(kcpCrd.Generation, 10)
+	runtimeCRDGeneration := strconv.FormatInt(runtimeCrd.Generation, 10)
+	return kyma.Annotations[kcpAnnotation] != latestGeneration ||
+		kyma.Annotations[skrAnnotation] != runtimeCRDGeneration
+}
+
+func updateKymaAnnotations(kyma *v1beta2.Kyma, crd *v1extensions.CustomResourceDefinition, crdType CrdType) {
+	if kyma.Annotations == nil {
+		kyma.Annotations = make(map[string]string)
+	}
+	annotation := getAnnotation(crd, crdType)
+	kyma.Annotations[annotation] = strconv.FormatInt(crd.Generation, 10)
+}
+
+func getAnnotation(crd *v1extensions.CustomResourceDefinition, crdType CrdType) string {
+	return fmt.Sprintf("%s-%s-crd-generation", strings.ToLower(crd.Spec.Names.Kind), strings.ToLower(string(crdType)))
+}
+
+func SyncCrdsAndUpdateKymaAnnotations(ctx context.Context, kyma *v1beta2.Kyma,
+	runtimeClient Client, controlPlaneClient Client) (bool, error) {
+	kcpKymaCrd, skrKymaCrd, err := fetchCrds(ctx, controlPlaneClient, runtimeClient, v1beta2.KymaKind.Plural())
```

Comment:
you can extract line 78 to 97 as a function, reuse it to fetchCrds and updateKymAnnotation for kyma and moduletemplate.

line 99 to 119 is basically just copy-paste which you should avoid.